### PR TITLE
[Metrics][Spec Decoding] Add per-request acceptance rate Prometheus histogram

### PR DESCRIPTION
**Pull request by jeffye-dev**
## Purpose

This PR adds request-level observability for speculative decoding acceptance.

vLLM already exposes aggregate speculative decoding counters such as draft counts and accepted-token counts. Those counters are useful for overall speculative decoding efficiency, but they are aggregated across scheduler
iterations and do not directly show the distribution of acceptance rates across finished requests.
This PR adds a finished-request-level Prometheus histogram vllm:request_spec_decode_acceptance_rate to inspect request-level acceptance behavior in production.

The motivation is based on the observation: speculative decoding metrics can be useful signals for detecting output-quality anomalies in production inference. In particular:

- Extremely low speculative acceptance can correlate with garbled or rare-character outputs, where draft tokens are almost entirely rejected by the target model.
- Extremely high speculative acceptance can correlate with repetitive outputs, where a corrupted inference state may push generation into a high-confidence loop.

This turns speculative decoding metrics from only a performance signal into a lightweight production monitoring signal for request-level output-quality anomalies.

## Test Plan

Unit tests cover:

- Accumulating request-level speculative draft and accepted token counts through
  `IterationStats`.
- `SpecDecodeRequestStats.merge()`.
- Observing `vllm:request_spec_decode_acceptance_rate` for a finished request.

## Test Result

Local tests are passed

### CODE REVIEW

**Review by markmc — CHANGES_REQUESTED** [review]
Thanks for your interest in this!

Our design philosophy is that we expose raw counters, and let operators compose derived metrics/alerts in their monitoring stack. We don't wish to bake threshold logic and alerting policy into the engine itself.

Instead, you can add e.g. `vllm:spec_decode_request_acceptance_rate (Histogram)`

observed once per finished request that had spec decode tokens. This lets operators write whatever threshold alerts they want:

```
  # Alert: >5% of requests have acceptance rate below 0.1
  sum(rate(vllm:spec_decode_request_acceptance_rate_bucket{le="0.1"}[5m]))
  /
  sum(rate(vllm:spec_decode_request_acceptance_rate_count[5m]))
  > 0.05

  # Alert: any requests with suspiciously high acceptance (repetition)
  sum(rate(vllm:spec_decode_request_acceptance_rate_bucket{le="0.96"}[5m]))
  /
  sum(rate(vllm:spec_decode_request_acceptance_rate_count[5m]))
  < 0.95
```

**Comment by jeffye-dev** [comment]
> Thanks for your interest in this!
> 
> Our design philosophy is that we expose raw counters, and let operators compose derived metrics/alerts in their monitoring stack. We don't wish to bake threshold logic and alerting policy into the engine itself.
> 
> Instead, you can add e.g. `vllm:spec_decode_request_acceptance_rate (Histogram)`
> 
> observed once per finished request that had spec decode tokens. This lets operators write whatever threshold alerts they want:
> 
> ```
>   # Alert: >5% of requests have acceptance rate below 0.1
>   sum(rate(vllm:spec_decode_request_acceptance_rate_bucket{le="0.1"}[5m]))
>   /
>   sum(rate(vllm:spec_decode_request_acceptance_rate_count[5m]))
>   > 0.05
> 
>   # Alert: any requests with suspiciously high acceptance (repetition)
>   sum(rate(vllm:spec_decode_request_acceptance_rate_bucket{le="0.96"}[5m]))
>   /
>   sum(rate(vllm:spec_decode_request_acceptance_rate_count[5m]))
>   < 0.95
> ```

thanks @markmc , I have updated the PR according to your comment. Now the metric is changed to "vllm:spec_decode_request_acceptance_rate" as histogram metric.

**Comment by markmc** [comment]
/cc @benchislett @luccafong @MatthewBonanni


Please update the PR description with the latest proposed metrics design ...

I'm still a bit unconvinced by the motivation:

> This PR adds request-level observability for speculative decoding acceptance anomalies.
> 
> The motivation is based on the observation: speculative decoding metrics can be useful signals for detecting output-quality anomalies in production inference. In particular:
> 
> * Extremely low speculative acceptance can correlate with garbled or rare-character outputs, where draft tokens are almost entirely rejected by the target model.
> * Extremely high speculative acceptance can correlate with repetitive outputs, where a corrupted inference state may push generation into a high-confidence loop.
> 
> vLLM already exposes step-level speculative decoding metrics, but those aggregate counters are not sufficient to identify whether an individual finished request had an abnormal acceptance pattern. This PR adds request-level accounting so operators can monitor such cases directly.

So the operator can find out if there is a certain class of requests that have a lower acceptance rate? In what was is this more actionable in practise than the existing aggregate metrics?

File: vllm/v1/core/sched/scheduler.py [hunk 1]
```diff
@@ -1414,26 +1414,33 @@ def update_from_output(
...
+            num_invalid_spec_tokens = 0
             if scheduled_spec_token_ids and generated_token_ids:
                 num_draft_tokens = len(scheduled_spec_token_ids)
-                num_accepted = len(generated_token_ids) - 1
-                num_rejected = num_draft_tokens - num_accepted
+                num_accepted_tokens = len(generated_token_ids) - 1
+                num_rejected_tokens = num_draft_tokens - num_accepted_tokens
+                if scheduler_output.num_invalid_spec_tokens:
+                    num_invalid_spec_tokens = (
+                        scheduler_output.num_invalid_spec_tokens.get(req_id, 0)
```
Comment (markmc) [inline vllm/v1/core/sched/scheduler.py]: This lookup was in `make_stats()` to avoid it when `--disable-log-stats` is used

File: vllm/v1/core/sched/scheduler.py [hunk 2]
```diff
@@ -1565,6 +1572,9 @@ def update_from_output(
                         trace_headers=request.trace_headers,
                         routed_experts=routed_experts,
                         num_nans_in_logits=request.num_nans_in_logits,
+                        num_draft_tokens=num_draft_tokens,
+                        num_accepted_tokens=num_accepted_tokens,
+                        num_invalid_spec_tokens=num_invalid_spec_tokens,
```
Comment (markmc) [inline vllm/v1/core/sched/scheduler.py]: I don't think you're using this field in practise

File: vllm/v1/metrics/stats.py [hunk 3]
```diff
@@ -218,6 +218,8 @@ class RequestStateStats:
 
     # Track if this request is corrupted (NaNs in logits)
     is_corrupted: bool = False
+    num_draft_tokens: int = 0
+    num_accepted_tokens: int = 0
```
Comment (markmc) [inline vllm/v1/metrics/stats.py]: I think it might work out nice to add something like this to `vllm.v1.spec_decode.metrics`

```
  class SpecDecodeRequestStats(NamedTuple):
      num_accepted_tokens: int
      num_draft_tokens: int

      @property
      def acceptance_rate(self) -> float:
          return self.num_accepted_tokens / self.num_draft_tokens
```

and use for `EngineCoreOutput`, `spec_decoding.observe_finished_request()`, `RequestStateStats`, `FinishedRequestStats`

File: vllm/v1/spec_decode/metrics.py [hunk 4]
```diff
@@ -148,9 +149,35 @@ def __init__(
         per_engine_labelvalues: dict[int, list[object]],
     ):
         self.spec_decoding_enabled = speculative_config is not None
+
         if not self.spec_decoding_enabled:
             return
 
+        histogram_request_acceptance_rate = self._histogram_cls(
+            name="vllm:spec_decode_request_acceptance_rate",
```
Comment (markmc) [inline vllm/v1/spec_decode/metrics.py]: Following the other per-request histogram naming conventions, this should be `vllm:request_spec_decode_acceptance_rate`

**Review by markmc — CHANGES_REQUESTED** [review]

**Comment by benchislett** [comment]
I'm okay with this as long as we clean it up. Ideally, I would like to have an option to return the per-request metrics to the client in the usage field. But that can be a follow-up or separate project. For monitoring, I'm okay with having a request-level histogram.

**Comment by jeffye-dev** [comment]
> /cc @benchislett @luccafong @MatthewBonanni
> 
> Please update the PR description with the latest proposed metrics design ...
> 
> I'm still a bit unconvinced by the motivation:
> 
> > This PR adds request-level observability for speculative decoding acceptance anomalies.
> > The motivation is based on the observation: speculative decoding metrics can be useful signals for detecting output-quality anomalies in production inference. In particular:
> > 
> > * Extremely low speculative acceptance can correlate with garbled or rare-character outputs, where draft tokens are almost entirely rejected by the target model.
> > * Extremely high speculative acceptance can correlate with repetitive outputs, where a corrupted inference state may push generation into a high-confidence loop.
> > 
> > vLLM already exposes step-level speculative decoding metrics, but those aggregate counters are not sufficient to identify whether an individual finished request had an abnormal acceptance pattern. This PR adds request-level accounting so operators can monitor such cases directly.
> 
> So the operator can find out if there is a certain class of requests that have a lower acceptance rate? In what was is this more actionable in practise than the existing aggregate metrics?

Thanks for this comment, I've updated the PR description.

This metric is very useful to detect timely the abnormal outputs in production environments, such as repetitive outputs or garbled text. This problems are uncommon, but when they do occur, they can be very serious.

The challenge is that detecting abnormal outputs in large-scale inference environments is quite difficult. For example, in our environment, we deploy a PD-disaggregated architecture based on NIXL. When abnormal conditions such as timeouts occur in KV cache communication between the prefill and decode instances, the KV cache on the decode instance can become corrupted, which in turn leads to abnormal outputs.
For long time, it was very hard to detect and had troubled us, until we discovered that the speculative decoding acceptance rate is a very clear signal of output anomalies. I also referenced to this blog where you can get more details: https://z.ai/blog/scaling-pain

File: vllm/v1/core/sched/scheduler.py (hunk 1 above)
Comment (jeffye-dev) [reply vllm/v1/core/sched/scheduler.py]: Thanks for the suggestion, I've done refactoring the codes: a new class SpecDecodeRequestStats is added and related logics are changed accordingly.

File: vllm/v1/core/sched/scheduler.py [hunk 5]
```diff
@@ -1414,10 +1414,21 @@ def update_from_output(
...
+            spec_decode_request_stats: SpecDecodeRequestStats | None = None
             if scheduled_spec_token_ids and generated_token_ids:
                 num_draft_tokens = len(scheduled_spec_token_ids)
                 num_accepted = len(generated_token_ids) - 1
                 num_rejected = num_draft_tokens - num_accepted
+                if self.log_stats:
+                    effective_num_draft_tokens = num_draft_tokens
+                    if scheduler_output.num_invalid_spec_tokens:
+                        effective_num_draft_tokens -= (
+                            scheduler_output.num_invalid_spec_tokens.get(req_id, 0)
```
Comment (markmc) [inline vllm/v1/core/sched/scheduler.py]: This might be a non-trivial lookup with many concurrent requests

Looks like it's easy to have `make_spec_decoding_stats()` return `spec_decode_request_stats`, which means we only need to do it once?

File: vllm/v1/metrics/stats.py [hunk 6]
```diff
@@ -371,6 +373,13 @@ def update_from_output(
             req_stats.first_token_latency = first_token_latency
 
         req_stats.num_generation_tokens += num_new_generation_tokens
+        if output.spec_decode_stats is not None:
```
Comment (markmc) [inline vllm/v1/metrics/stats.py:392]: A question of taste perhaps, but maybe:

```
if output.spec_decode_stats is not None:
    req_stats.spec_decode_stats = output.spec_decode_stats.merge(req_stats.spec_decode_stats)
```

moving the `is None` check into `merge()`

**Review by markmc — CHANGES_REQUESTED** [review]
Main thing to address is the failing DCO check - you need to add `Signed-off-by`

While you're modifying the commit message though, would be nice to line wrap it (at 72 chars is conventional)

File: vllm/v1/metrics/stats.py (hunk 6 above)
Comment (jeffye-dev) [reply vllm/v1/metrics/stats.py:392]: updated

File: vllm/v1/core/sched/scheduler.py (hunk 5 above)
Comment (markmc) [reply vllm/v1/core/sched/scheduler.py]: The invalid tokens lookup is still happening twice in the latest version, I was proposing:

```
    def make_spec_decoding_stats(
        self,
        spec_decoding_stats: SpecDecodingStats | None,
        num_draft_tokens: int,
        num_accepted_tokens: int,
        num_invalid_spec_tokens: dict[str, int] | None,
        request_id: str,
    ) -> tuple[SpecDecodingStats | None, SpecDecodeRequestStats | None]:
        if not self.log_stats or not num_draft_tokens:
            return None, None
        if spec_decoding_stats is None:
            spec_decoding_stats = SpecDecodingStats.new(self.num_spec_tokens)
        if num_invalid_spec_tokens:
            num_draft_tokens -= num_invalid_spec_tokens.get(request_id, 0)
        spec_decoding_stats.observe_draft(
            num_draft_tokens=num_draft_tokens,
            num_accepted_tokens=num_accepted_tokens,
        )
        return (
            spec_decoding_stats,
            SpecDecodeRequestStats(
                num_draft_tokens=num_draft_tokens,
                num_accepted_tokens=num_accepted_tokens,
            ),
        )

```

Note - I have removed the `max(0, num_draft)` clamp - let's not hide real bugs like this. Add an assertion if you believe such bugs are lurking

Comment (jeffye-dev) [reply vllm/v1/core/sched/scheduler.py]: thanks for pointing this, I've updated the codes.

**Review by markmc — APPROVED** [review]

**Comment by markmc** [comment]
pre-commit issue should be fixed by #45374

**Comment by jeffye-dev** [comment]
The buildkite/ci/pr failure seemed not caused by this pr.
